File: Data Structures & Algorithms/eating-bananas/submission-2.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:

        lp = 1
        rp = max(piles)
        result = rp

        while lp <= rp:

            mid = (lp + rp) // 2
            totalH = 0

            for bananas in piles:
                totalH += math.ceil(bananas / mid)

            if totalH <= h:
                result = mid
                rp = mid - 1
            else:
                lp = mid + 1

        return result

File: Data Structures & Algorithms/eating-bananas/submission-2.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = 1;
        int currFasted = 0;

        for (int pile: piles) right = Math.max(right, pile);

        while (left <= right){

            int middle = left + ((right-left)/2);

            int duration = 0;
            for (int pile: piles){
                duration += Math.ceil((double) pile / middle);
            }

            if (duration <= h){
                currFasted = middle;
                right = middle - 1;
            } else {
                left = middle + 1;
            }
        }

        return currFasted;
        
        
    }
}
